text/trace/0192-messaging-semantic-conventions-spec.md

```diff
@@ -0,0 +1,659 @@
+                   .              |
+Create             .           Process
+  |                v              ^
+  v        +--------------+       |
+Publish -> | INTERMEDIARY | -> Receive
+           +--------------+
+```
+
+1. The producer creates a message.
+2. The producer publishes the message to an intermediary.
+3. The consumer receives the message from an intermediary.
+4. The consumer processes the message.
+5. The consumer settles the message by notifying the intermediary that the
+   message was processed. In some cases (fire-and-forget), the settlement stage
+   does not exist.
+
+The semantic conventions described below define how to model those stages in
+traces, how to propagate context, and how to enrich traces with attributes.
+
+### Context propagation
+
+A message may pass many different components and layers in one or more
+intermediaries when it is propagated from the producer to the consumer. It
+cannot be assumed, and in many cases it is not even desired, that all those
+components and layers are instrumented and propagate context according to
+OpenTelemetry requirements.
+
+A _creation context_ allows correlating the producer with the consumers of a
+message, regardless of intermediary instrumentation. The creation context is
+created by the producer and must be propagated to the consumers. It must not be
+altered by intermediaries.  This context helps to model dependencies between
+producers and consumers, regardless of the underlying messaging transport
+mechanism and its instrumentation.
+
+Instrumentors are required to instrument producer and consumer applications
+so that context is attached to messages and extracted from messages in a
+coordinated way. Future versions of these conventions might recommend [context propagation according to certain industry standards](#standards-for-context-propagation).
+
+> A producer SHOULD attach a creation context to each message. The creation context
+> SHOULD be attached in a way so that it is not possible to be changed by intermediaries.
+
+### Trace structure
+
+#### Producer
+
+Producers are responsible for attaching a creation context to a message.
+Subsequent consumers will use this context to link consumer traces to producer
+traces. Ideally, each message gets a unique and distinct creation context
+assigned. However, as a context must refer to a span this would require the
+creation of a distinct span for each message, which is not feasible in all
+scenarios. In certain batching scenarios where many messages are created and
+published in large batches, creating a span for every single message would
+obfuscate traces and is not desirable. Thus having a unique and distinct
+context per message is recommended, but not required.
+
+For each producer scenario, a "Publish" span needs to be created. This span
+measures the duration of the call or operation that provides messages for
+sending or publishing to an intermediary. This call or operation (and the
```

Comment:
My interpretation is that it's "vague" intentionally. In my view it can mean/cover all of the scenarios you brought. The problem is that it can't enforce anything and since it changes between systems + what instrumentations can do I'm not sure if we should be highly specific here.

Reply:
I made a suggestion to list the scenarios.